File: go/mechanisms/svm/upto/facilitator/channel_storage.go
package facilitator

import (
	"context"
	"sync"
	"time"
)

// ChannelRecord holds the payment-channel facts rent cleanup needs and the
// channel account itself cannot provide: the distribution preimage (payTo),
// the token program, the abandon-policy timestamps, and the network.
// Payer, payee, mint, openSlot, and status are always read live before acting.
//
// Written on deposit settle (before broadcast) and on claim settle; deleted
// once the PDA is gone.
type ChannelRecord struct {
	ChannelID string
	// PayTo is the distribution recipient sealed at open.
	PayTo        string
	TokenProgram string
	// FirstSeenAt is when the facilitator first stored this channel.
	FirstSeenAt time.Time
	// ExpiresAt is the client voucher expiry (Unix seconds). It never shrinks
	// on a later upsert.
	ExpiresAt int64
	Network   string
}

// ChannelStorage persists the channels a facilitator sponsors rent for.
// Implementations must be safe for concurrent use; inject a durable one for
// multi-process facilitators.
type ChannelStorage interface {
	Get(ctx context.Context, channelID string) (*ChannelRecord, error)
	// List returns every stored record, in any order. The rent cleanup manager
	// sorts by channel ID before scanning, so implementations do not have to.
	List(ctx context.Context) ([]ChannelRecord, error)
	Upsert(ctx context.Context, record ChannelRecord) error
	Delete(ctx context.Context, channelID string) error
}

// InMemoryChannelStorage is the default ChannelStorage. It keeps the earliest
// FirstSeenAt and the latest ExpiresAt across upserts of the same channel.
type InMemoryChannelStorage struct {
	mu       sync.RWMutex
	channels map[string]ChannelRecord
}

// NewInMemoryChannelStorage creates an empty in-memory channel store.
func NewInMemoryChannelStorage() *InMemoryChannelStorage {
	return &InMemoryChannelStorage{channels: make(map[string]ChannelRecord)}
}

// Get returns a stored channel, or nil when the channel is not tracked.
func (s *InMemoryChannelStorage) Get(_ context.Context, channelID string) (*ChannelRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	record, ok := s.channels[channelID]
	if !ok {
		return nil, nil
	}
	return &record, nil
}

// List returns every stored channel record.
func (s *InMemoryChannelStorage) List(_ context.Context) ([]ChannelRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	records := make([]ChannelRecord, 0, len(s.channels))
	for _, record := range s.channels {
		records = append(records, record)
	}
	return records, nil
}

// Upsert inserts or replaces a record, preserving the original FirstSeenAt and
// the longest ExpiresAt seen for the channel.
func (s *InMemoryChannelStorage) Upsert(_ context.Context, record ChannelRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if existing, ok := s.channels[record.ChannelID]; ok {
		record.FirstSeenAt = existing.FirstSeenAt
		if existing.ExpiresAt > record.ExpiresAt {
			record.ExpiresAt = existing.ExpiresAt
		}
	}
	s.channels[record.ChannelID] = record
	return nil
}

// Delete removes a channel from storage.
func (s *InMemoryChannelStorage) Delete(_ context.Context, channelID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.channels, channelID)
	return nil
}

File: typescript/packages/mechanisms/svm/src/upto/facilitator/channelStorage.ts
import type { Network } from "@x402/core/types";

/**
 * Stored payment-channel facts used by facilitator rent cleanup.
 *
 * Only stores what the channel account refetch does not provide:
 * `payTo` (distribution preimage), `tokenProgram`, abandon-policy
 * `firstSeenAt`, voucher `expiresAt`, and `network`. Payer/payee/mint/
 * openSlot/status are read live before acting.
 *
 * Written on deposit (pre-broadcast) and claim settle; deleted when the PDA is gone.
 */
export interface UptoChannelRecord {
  channelId: string;
  /** Distribution recipient sealed at open (`requirements.payTo`). */
  payTo: string;
  tokenProgram: string;
  /** Wall-clock ms when the facilitator first stored this channel. */
  firstSeenAt: number;
  /** Client voucher expiry (Unix seconds). Never shrinks on later upserts. */
  expiresAt: number;
  network: Network;
}

/** Pluggable storage of channels the facilitator sponsors rent for. */
export interface UptoChannelStorage {
  get(channelId: string): Promise<UptoChannelRecord | undefined>;
  /**
   * Every stored record, in any order. The rent cleanup manager sorts by
   * channel id before scanning, so implementations do not have to.
   */
  list(): Promise<UptoChannelRecord[]>;
  upsert(record: UptoChannelRecord): Promise<void>;
  delete(channelId: string): Promise<void>;
}

/**
 * In-memory {@link UptoChannelStorage}. Preserves `firstSeenAt` and the
 * maximum `expiresAt` across upserts of the same `channelId`.
 */
export class InMemoryUptoChannelStorage implements UptoChannelStorage {
  private readonly channels = new Map<string, UptoChannelRecord>();

  /**
   * Look up a single stored channel.
   *
   * @param channelId - Channel PDA
   * @returns Stored record, or undefined when absent
   */
  async get(channelId: string): Promise<UptoChannelRecord | undefined> {
    return this.channels.get(channelId);
  }

  /**
   * List every stored channel record.
   *
   * @returns All stored channel records
   */
  async list(): Promise<UptoChannelRecord[]> {
    return [...this.channels.values()];
  }

  /**
   * Insert or replace a record. Keeps the earlier `firstSeenAt` and the
   * later `expiresAt` when the channel was already stored.
   *
   * @param record - Full channel storage record
   */
  async upsert(record: UptoChannelRecord): Promise<void> {
    const existing = this.channels.get(record.channelId);
    this.channels.set(record.channelId, {
      ...record,
      firstSeenAt: existing?.firstSeenAt ?? record.firstSeenAt,
      expiresAt: existing ? Math.max(existing.expiresAt, record.expiresAt) : record.expiresAt,
    });
  }

  /**
   * Remove a channel from storage.
   *
   * @param channelId - Channel PDA to remove
   */
  async delete(channelId: string): Promise<void> {
    this.channels.delete(channelId);
  }
}
